### ISSUE TITLE
Haddock paragraph whitespace

### ISSUE THREAD

**Issue by pbrisbin** [freckle/guides#20]
I'm going to add some Haddock guidelines to this repo. AFAIK, what I would add is not controversial and I expect the PR to go by with minimal comments, except for one very specific thing: the whitespace around block content, specifically the empty line before the function itself...

So I'm bringing that up here first, so I can include whatever we decide in that eventual PR. I have my own preference, and I've included arguments for it below, but I'm happy to go with whatever the team votes -- assuming it comes to that.

### A: White-space around block content

```hs
-- | Return all Assessors for a @'SchoolAdmin'@
--
-- An __Assessor__ is any admin that has assigned at least one district
-- assessment to at least one student accessible to the signed-in admin. This
-- always includes the signed-in admin to avoid a race between school-api (where
-- district assessments are created and fetched) and here (where admins are
-- fetched).
--
getSchoolAdminsAssessorsR :: Handler [SchoolAdminAssessor]
getSchoolAdminsAssessorsR = do
```

This is what I naturally do without thinking. I like the extra breathing room around all block content (for me, this extends to other formats such as markdown). I think it makes things objectively more readable, but I could just be bias. This rule is also simple to describe, without any special cases, which is usually a benefit in Guides.

I also think there is historical precedent that we prefer:

```hs
{-# LANGUAGE -}

module Foo where
```

And not

```hs
{-# LANGUAGE -}
module Foo where
```

So we like visual separation generally. This feels like a similar decision to me.

### B: White-space around block content, except last line

```hs
-- | Return all Assessors for a @'SchoolAdmin'@
--
-- An __Assessor__ is any admin that has assigned at least one district
-- assessment to at least one student accessible to the signed-in admin. This
-- always includes the signed-in admin to avoid a race between school-api (where
-- district assessments are created and fetched) and here (where admins are
-- fetched).
getSchoolAdminsAssessorsR :: Handler [SchoolAdminAssessor]
getSchoolAdminsAssessorsR = do
```

This seems to be what @cdparks and @eborden do naturally without thinking. I will let them add any objective readability points about this themselves, if there are any. The main value to this style is that 2 of the team members do it naturally (vs 1 who does the other).

### C: White-space around block content, except first/last line

```hs
-- | Return all Assessors for a @'SchoolAdmin'@
-- An __Assessor__ is any admin that has assigned at least one district
-- assessment to at least one student accessible to the signed-in admin. This
-- always includes the signed-in admin to avoid a race between school-api (where
-- district assessments are created and fetched) and here (where admins are
-- fetched).
getSchoolAdminsAssessorsR :: Handler [SchoolAdminAssessor]
getSchoolAdminsAssessorsR = do
```

~~Presented without comment.~~ Feels entirely too cramped to me :)

### D: One big paragraph

```hs
-- | Return all Assessors for a @'SchoolAdmin'@. An __Assessor__ is any admin
-- that has assigned at least one district assessment to at least one student
-- accessible to the signed-in admin. This always includes the signed-in admin
-- to avoid a race between school-api (where district assessments are created
-- and fetched) and here (where admins are fetched).
getSchoolAdminsAssessorsR :: Handler [SchoolAdminAssessor]
getSchoolAdminsAssessorsR = do
```

I don't think anyone would advocate for this. Just like commit messages, Haddocks are better with a distinct Summary and Body.

### E: One big pragraph, hanging indent

```hs
-- | Return all Assessors for a @'SchoolAdmin'@. An __Assessor__ is any admin
--   that has assigned at least one district assessment to at least one student
--   accessible to the signed-in admin. This always includes the signed-in admin
--   to avoid a race between school-api (where district assessments are created
--   and fetched) and here (where admins are fetched).
getSchoolAdminsAssessorsR :: Handler [SchoolAdminAssessor]
getSchoolAdminsAssessorsR = do
```

Again, I don't think anyone wants this, but I've included it for completeness.

Also worth noting: unlike (A) vs (B) vs (C), those vs (D) and (E) will actually change how the rendered documentation looks.

There are other mix-and-match variations between these that I've excluded. Please add them if you want to advocate for them as an option.

## Frequencies

Because (a) I write a lot of haddocks and (b) people have accommodated me when I've requested a certain Haddock format in Code Review, (A) is currently the most frequent style across the codebase\*

But I'm not sure that's the only factor. I continue to comment on (B), asking it to be (A) in code review so I think if I stopped doing that for a little while (B) would catch up.

\* Caveats:

1. This is real hard to measure, so I just did a very rough inexact `grep` and visual ballpark count
2. This is only if we exclude anything that's not a properly formatted Haddock at all

## Discussion vs Voting

Please discuss if you want, of course. But if you want to just vote, throw a :+1: on here for (A) and a :-1: for (B). I will add a vote for myself.

**Edited to say "Summary", not "Title"**

**Comment by jdreaver** [comment]
I'm :+1:  with the space rules in option A but, does this mean we *must* have a title line? What about this:

```haskell
-- | This is a haddocks comment that I am making up for the purpose of demonstration
-- and it is kinda long. Not too long though, since it is just two sentences.
--
foo :: Int
foo = 1
```

Obviously a rule of thumb for writing in general is to break up long paragraphs, but certainly this paragraph is not too long. I'm all for option A, but if option A _enforces_ having a title then I think I'm :-1: 

(Note: option A doesn't talk about the Title, option D does, which is why I got this impression):

> Haddocks are better with a distinct Title and Body.

**Comment by pbrisbin** [comment]
Your example would not be allowed, IMO. What you describe is an omitted mix-match of (D) and (A).

Some other rules I plan to propose are:

- Haddocks must have a Summary and *optional* Body
- Summaries have some rules that are inspired by [good commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html), in this case it must be a **single, short, and un-punctuated** statement

Only if there is a Body, do the rules in this Issue apply.

So your example would either be:

```hs
-- | Something like this
foo :: Int
```

if it truly is sufficiently simple. Otherwise, you have to jump to

```hs
-- | Something shorter
--
-- With whatever you were planning above being here instead. This is a haddocks
-- comment that I am making up for the purpose of demonstration and it is kinda
-- long. Not too long though, since it is just two sentences.
--
foo :: Int
foo = 1
```

I know you probably want the freedom to go with "but it's just two sentences", but:

- Not enforcing this really damages the goal of consistency and scannability
- Enforcing this will foster _more_ documentation, because once you've made the jump from a Summary to a Summary+Body, you are naturally willing to write more in the Body -- this how it works in commit messages

To recap:

1. Option (A) does not enforce that there is a Body
2. Other rules would add restrictions on Summary length and formatting
3. If what you want to write can't fit in (2), you should have a Body

If (2) is controversial, maybe we can open a separate Issue and elicit a separate discussion?

**Edited to say "Summary", not "Title"**

**Comment by eborden** [comment]
I have no skin in this game. You seem passionate about the first option, so I'm down for it.

**Comment by tfausak** [comment]
Just jumping in to say that option A reminds me of [Python's docstring conventions](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings):

> Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.

So there's precedent for such a guideline 🙂

**Comment by jdreaver** [comment]
Yeah it reminds me of Python too. I can get onboard with a title + body. I tend to do title + body anyway for much longer Haddock comments, so stomping out the inconsistency for shorter ones should be fine.

**Comment by pbrisbin** [comment]
I guess now's the time to decide then:

- Title + Body (what I did here, for some reason?)
- Summary + Body (python, git commits)
- Subject + Body (email)

**Edit: Summary won**.

**Comment by tfausak** [comment]
For what it's worth, I like option B. The extra blank line before the type signature looks weird to me. But not weird enough that I would actually fight it. And to continue cribbing from Python, every docstring ends with that triple quote on its own line.

**Comment by jdreaver** [comment]
Ah I thought Title meant Summary. If we are being explicit then Summary seems like the right word/style.

**Comment by pbrisbin** [comment]
Done. Edited existing stuff to use "Summary" consistently.

**Comment by cdparks** [comment]
I used to do B, but @pbrisbin brainwashed me, so I've been doing A. I don't really prefer one or the other, but if A's already got the numbers, cool. Hard :+1: on Summary + Body, just like commit messages

**Comment by eborden** [comment]
![](https://media.giphy.com/media/YdVSZTDageO2I/giphy.gif)

**Comment by 5outh** [comment]
I'm :+1: on A!

**Comment by pbrisbin** [comment]
Closing this issue. My PR will include (A) as well as the other rules hinted at in this thread.

Don't let it being closed stop the GIFs though.

![](https://media1.tenor.com/images/fcc5d05c5899c7506dded7956dfa8953/tenor.gif?itemid=10178892)

**Comment by tfausak** [comment]
![Even though these rules don't apply to me, I'll learn to like the new way.](https://i.makeagif.com/media/5-31-2014/jNAuZf.gif)
